source/concepts.tex

```diff
@@ -0,0 +1,1249 @@
+is publicly and unambiguously derived from \tcode{Base}, or \tcode{Derived} and
+\tcode{Base} are the same class type ignoring cv-qualifiers.
+\end{note}
+\end{itemdescr}
+
+\rSec2[concepts.lib.corelang.convertibleto]{Concept \tcode{ConvertibleTo}}
+
+\pnum
+The \tcode{ConvertibleTo} concept requires an expression of a particular type
+and value category to be both implicitly and explicitly convertible to some other
+type. The implicit and explicit conversions must produce equal results.
+
+\indexlibrary{\idxcode{ConvertibleTo}}%
+\begin{itemdecl}
+template <class From, class To>
+concept ConvertibleTo = is_convertible_v<From, To> &&
+  requires(From (&f)()) { static_cast<To>(f()); };
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+Let \tcode{test} be the invented function:
+\begin{codeblock}
+To test(From (&f)()) {
+  return f();
+}
+\end{codeblock}
+and let \tcode{f} be a function with no arguments and return type \tcode{From}
+such that \tcode{f()} is equality preserving.
+\tcode{ConvertibleTo<From, To>} is satisfied only if:
+
+\begin{itemize}
+\item
+\tcode{To} is not an object or reference-to-object type, or
+\tcode{static_cast<To>(f())} is equal to \tcode{test(f)}.
+
+\item
+\tcode{From} is not a reference-to-object type, or
+
+\begin{itemize}
+\item
+If \tcode{From} is an rvalue reference to a non const-qualified type, the resulting
+state of the object referenced by \tcode{f()} after either above expression is
+valid but unspecified\iref{lib.types.movedfrom}.
+
+\item
+Otherwise, the object referred to by \tcode{f()} is not modified by either above expression.
+\end{itemize}
+
+\end{itemize}
+
+\pnum
+There need be no subsumption relationship between \tcode{ConvertibleTo<From, To>}
+and \tcode{is_\-convert\-ible_v\brk{}<From,\brk{} To>}.
+\end{itemdescr}
+
+
+\rSec2[concepts.lib.corelang.commonref]{Concept \tcode{CommonReference}}
+
+\pnum
+For two types \tcode{T} and \tcode{U}, if \tcode{common_reference_t<T, U>}
+is well-formed and denotes a type \tcode{C} such that both
+\tcode{ConvertibleTo<T, C>} and \tcode{ConvertibleTo<U, C>} are
+satisfied, then \tcode{T} and \tcode{U} share a \term{common reference
+type}, \tcode{C}.
+\begin{note}
+\tcode{C} could be the same as \tcode{T}, or \tcode{U}, or it could be a different
+type. \tcode{C} may be a reference type.
+\end{note}
+
+\indexlibrary{\idxcode{CommonReference}}%
+\begin{itemdecl}
+template <class T, class U>
+concept CommonReference =
+  Same<common_reference_t<T, U>, common_reference_t<U, T>> &&
+  ConvertibleTo<T, common_reference_t<T, U>> &&
+  ConvertibleTo<U, common_reference_t<T, U>>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+Let \tcode{C} be \tcode{common_reference_t<T, U>}. Let \tcode{t} be a
+function whose return type is \tcode{T}, and let \tcode{u} be a function
+whose return type is \tcode{U}. \tcode{CommonReference<T, U>} is satisfied
+only if:
+\begin{itemize}
+\item \tcode{C(t())} equals \tcode{C(t())} if and only if \tcode{t()} is an
```

Comment:
The "if and only if" feels wrong here; it seems to say that C(t()) must have a value different from C(t()) if t() is not equality preserving. That seems hard to do, since t() could modify some unknown global variable.

Reply:
Agreed with an LWG issue.